src/core/net/ip6.cpp

```diff
@@ -603,8 +614,303 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+        else
+        {
+            payloadLeft -= maxPayloadFragment;
+            payloadFragment = maxPayloadFragment;
+        }
+
+        offset = fragmentCnt * (maxPayloadFragment >> 3);
+        fragmentHeader.SetOffset(offset);
+
+        VerifyOrExit((fragment = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetLength(aMessage.GetOffset() + sizeof(fragmentHeader) + payloadFragment));
+
+        header.SetPayloadLength(payloadFragment + sizeof(fragmentHeader));
+        assertValue = fragment->Write(0, sizeof(header), &header);
+        assert(assertValue == sizeof(header));
+
+        SuccessOrExit(error = fragment->SetOffset(aMessage.GetOffset()));
+        assertValue = fragment->Write(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader);
+        assert(assertValue == sizeof(fragmentHeader));
+
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset() + static_cast<uint16_t>(offset << 3),
+                                     aMessage.GetOffset() + sizeof(fragmentHeader), payloadFragment,
+                                     *fragment) == static_cast<int>(payloadFragment),
+                     error = OT_ERROR_NO_BUFS);
+
+        EnqueueDatagram(*fragment);
+
+        fragmentCnt++;
+        fragment = NULL;
+
+        otLogInfoIp6("Fragment %d with %d bytes sent", fragmentCnt, payloadFragment);
+    }
+
+exit:
+    aMessage.Free();
+
+    if (error == OT_ERROR_NO_BUFS)
+    {
+        otLogWarnIp6("No buffer for Ip6 fragmentation");
+    }
+
+    if (error != OT_ERROR_NONE && fragment != NULL)
+    {
+        fragment->Free();
+    }
+
+    return error;
+}
+
+otError Ip6::HandleFragment(Message &aMessage, Netif *aNetif, MessageInfo &aMessageInfo, bool aFromNcpHost)
+{
+    otError        error = OT_ERROR_NONE;
+    Header         header, headerBuffer;
+    FragmentHeader fragmentHeader;
+    Message *      message         = NULL;
+    uint16_t       offset          = 0;
+    uint16_t       payloadFragment = 0;
+    int            assertValue     = 0;
+    bool           isFragmented    = true;
+
+    VerifyOrExit(aMessage.Read(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_PARSE);
+
+    VerifyOrExit(aMessage.Read(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader) == sizeof(fragmentHeader),
+                 error = OT_ERROR_PARSE);
+
+    if (fragmentHeader.GetOffset() == 0 && fragmentHeader.IsMoreFlagSet() == false)
+    {
+        isFragmented = false;
+        error        = aMessage.MoveOffset(sizeof(fragmentHeader));
+
+        ExitNow();
+    }
+
+    for (message = mReassemblyList.GetHead(); message; message = message->GetNext())
+    {
+        VerifyOrExit(aMessage.Read(0, sizeof(headerBuffer), &headerBuffer) == sizeof(headerBuffer),
+                     error = OT_ERROR_PARSE);
+        if (message->GetDatagramTag() == fragmentHeader.GetIdentification() &&
+            headerBuffer.GetSource().     operator==(header.GetSource()) &&
```

Comment:
Can we just have `headerBuffer.GetSource() == header.GetSource()`?

Reply:
yes, changed.